include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1233 @@
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+        other.comm_ = MPI_COMM_NULL;
+        other.size_ = 0;
+        other.rank_ = -1;
+    }
+
+    communicator& operator=(communicator&& other)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_dup(other.comm_, &this->comm_));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->local_rank_ = get_local_rank();
+        other.size_ = 0;
+        other.rank_ = -1;
+        return *this;
+    }
+
+    static MPI_Comm get_comm_world() { return MPI_COMM_WORLD; }
+
+    static std::shared_ptr<communicator> create_world()
+    {
+        return std::make_shared<communicator>(get_comm_world());
+    }
+
+    MPI_Comm get() const { return comm_; }
+
+    int size() const { return size_; }
+
+    int rank() const { return rank_; };
+
+    int local_rank() const { return local_rank_; };
+
+    bool compare(const MPI_Comm& other) const
+    {
+        int flag;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_compare(this->comm_, other, &flag));
+        return flag;
+    }
+
+    bool operator==(const communicator& rhs) { return compare(rhs.get()); }
+
+    ~communicator()
+    {
+        if (this->comm_ && this->comm_ != MPI_COMM_NULL) {
+            MPI_Comm_free(&this->comm_);
+        }
+    }
+
+private:
+    MPI_Comm comm_;
+    int size_{};
+    int rank_{};
+    int local_rank_{};
+
+    int get_my_rank()
+    {
+        int my_rank = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(comm_, &my_rank));
+        return my_rank;
+    }
+
+    int get_local_rank()
+    {
+        MPI_Comm local_comm;
+        int rank;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_split_type(
+            comm_, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &local_comm));
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(local_comm, &rank));
+        MPI_Comm_free(&local_comm);
+        return rank;
+    }
+
+    int get_num_ranks()
+    {
+        int size = 1;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_size(comm_, &size));
+        return size;
+    }
+};
+
+
+/**
+ * Get the rank in the communicator of the calling process.
+ *
+ * @param comm  the communicator
+ */
+inline double get_walltime() { return MPI_Wtime(); }
+
+
+/**
+ * This function is used to synchronize between the ranks of a given
+ * communicator.
+ *
+ * @param comm  the communicator
+ */
+inline void synchronize(const std::shared_ptr<communicator>& comm)
+{
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Barrier(comm->get()));
+}
+
+
+/**
+ * Allows a rank to wait on a particular request handle.
+ *
+ * @param req  The request to wait on.
+ * @param status  The status variable that can be queried.
+ */
+inline void wait(std::shared_ptr<request> req,
+                 std::shared_ptr<status> status = {})
+{
+    GKO_ASSERT_NO_MPI_ERRORS(
+        MPI_Wait(req->get(), status ? status->get() : MPI_STATUS_IGNORE));
```

Comment:
Even if the `req` object contains several `MPI_Request`s, you would be waiting on only the first one, I guess.